Add compstyle system

mattmc3 opened the pull request:
This PR relates to the discussion in issue #7. This is a working implementation of the `compstyle` system, which behaves similar to zsh's built-in `prompt` system. It includes:

- `compstyleinit`: the function to initialize comp styles
- `compstyle`: the function to set a completion style from $fpath
- `_compstyle` completions file to provide completions for the compstyle command itself
- four working `compstyle_xxx_setup` functions to get started with that are shameless copies of completions from Prezto, Oh-My-Zsh, and grml, as well as a unique zsh-utils one that takes a little from each with some good defaults.

This does not yet include:
- Good documentation

I'm not kidding when I say this - **shred this!** It's not ready to merge without robust feedback. I'm here to collaborate, so if there's something that is too clever, isn't well written, I forgot, or doesn't belong - let's work on it.

Some starting points:
- Do we want to include the shameless ripoffs from other projects? Are we okay with goofy 'catsup' names like `ohmy`, `prez`, and `grmlin`?
- Grml is GPL 2. Everything else is MIT. Is including code from it okay with you?
- I like using `##?` comments for usage documentation, but some find it too clever. If that's not okay, I can switch back to usage strings or HEREDOCs.
- I added a 'cleanup' system that nukes all ':completion:*' zstyles. I know we talked about not doing cleanup. Thoughts on this implementation?
- Others?

mattmc3 committed:
completion: add new compstyle system

mattmc3 committed:
completion: add common compstyle setups

belak commented on completion/functions/compstyle_zsh-utils_setup:
```diff
@@ -0,0 +1,93 @@
+  zstyle ':completion:*:default' list-colors ${(s.:.)LS_COLORS}
+  zstyle ':completion:*:default' list-prompt '%S%M matches%s'
+
+  # Use caching to make completion for commands such as dpkg and apt usable.
+  zstyle ':completion::complete:*' use-cache on
+  zstyle ':completion::complete:*' cache-path "${XDG_CACHE_HOME:-$HOME/.cache}/zsh/zcompcache"
```
I feel like the cache-related settings are fairly universally applicable, so we can leave them in the completion plugin and remove them from here.

EDIT: I just noticed that the compstyle cleanup removes all completion zstyles... so maybe there's a better way to do this? I'm open to suggestions. Maybe completion should also export the compstyle location as an environment variable to make sure it matches.

mattmc3 replied:
Oh yeah - forgot to mention that. When I added the cleanup functions I had to have a way to re-apply the cache zstyles, so I just added them to the setup file. I don't have to be so aggressive with the cleanup, but I was going for simplicity. Or, we could punt on cleanup functionality for now as originally discussed.

belak replied:
Maybe we could export a variable for the cache location and use that in `compstyle` to re-apply these 2 settings after cleanup?

mattmc3 replied:
Resolved in recent commit.

belak commented on completion/functions/compstyleinit:
```diff
@@ -0,0 +1,107 @@
+  fi
+
+  # cleanup any prior completion styles
+  if (( $+functions[compstyle_$1_cleanup] )); then
```
I'm not entirely sure how this should work, but if I'm reading this right, it uses the current compstyle name rather than the previous compstyle name for cleanup which seems slightly off.

mattmc3 replied:
My thinking here is that we provide a default `compstyle_cleanup`, but if a compstyle implementation provides its own `compstyle_xxx_cleanup` function, we call that instead. This was how I let the user override the cleanup functionality, but still be able to call the original `compstyle_cleanup` logic too if they wanted. Again, the cleanup part is probably not the most well thought out part of this.

belak replied:
Here's some clarification... this is how it currently works:

```sh
compstyle prez
# calls compstyle_prez_cleanup (if it exists), otherwise compstyle_cleanup
# calls compstyle_prez_setup

compstyle zsh-utils
# calls compstyle_zsh-utils_cleanup (if it exists), otherwise compstyle_cleanup
# calls compstyle_zsh-utils_setup
```

My expectation is more like this:
```sh
compstyle prez
# calls compstyle_cleanup (because there is no previously loaded compstyle)
# calls compstyle_prez_setup

compstyle zsh-utils
# calls compstyle_prez_cleanup (if it exists), otherwise compstyle_cleanup
# calls compstyle_zsh-utils_setup
```

The difference is that the cleanup function would be called when that compstyle is unloaded, rather than before it's loaded.

mattmc3 replied:
I've removed the cleanup functionality for now. Cleanup is complicated enough that it will require some thought to implement, and probably won't be worth it when opening a new shell is easy enough to ensure a fresh environment.

mattmc3 commented:
A few things to test (assuming the zsh-utlis completions plugin is loaded):

- Typing `compstyle -l` should list the 4 available compstyles
- Typing `compstyle <TAB>` (ie: compstyle followed by space followed by <kbd>TAB</kbd>) should show the 4 included compstyles
- Setting `compstyle zsh-utils` and then typing `ls<TAB>` (no spaces) should yield something like (with color):
```
 -- external command --
ls         lsappinfo  lsbom      lskq       lsm        lsmp       lsof       lsvfs
 -- alias --
ls
 -- parameter --
LS_COLORS
```
- Mis-spelling grep as `gerp` and hitting <kbd>TAB</kbd> to complete should yield spelling corrections (with color):
```
 -- corrections (errors: 1) --
gexpand  gexpr    grep
 -- original --
gerp
```

belak commented on completion/completion.plugin.zsh:
```diff
@@ -6,8 +6,12 @@ if [[ "$TERM" == 'dumb' ]]; then
 fi
 
+#
+# Config
```
I wrote this in contributing ages ago - https://github.com/belak/zsh-utils/blob/main/CONTRIBUTING.md#sections

I'm not sure it's accurate any more, but I would like to be consistent. I think that's why I left it under "Requirements" initially.

mattmc3 replied:
I'll be honest - I missed that part in the contributing guidelines. I have no strong preference here, so I can remove that.

belak commented on completion/functions/compstyle_grmlin_setup:
```diff
@@ -0,0 +1,189 @@
+### grml Project information:
+## Home:    https://grml.org
+## Authors: grml-team (grml.org)
+## License: GPL v2
```
As you mentioned in your comment, you weren't sure about this - unfortunately, the terms of the GPL don't allow including GPL licensed code in MIT licensed projects like this, so I think this one should be removed. It's a bit of a grey area because it's "just" configuration, but I think it would be easiest to drop this one for now.

mattmc3 replied:
It would be easy enough to include as its own separate GPL'ed plugin for grml fans. I'll drop it.

belak replied:
Yep - maybe we can include something later, I just don't want to think about the potential licensing headache just yet.

belak commented on completion/functions/compstyle_zsh-utils_setup:
```diff
@@ -0,0 +1,93 @@
+#!/bin/zsh
+
+function compstyle_zsh-utils_help () {
```
Help functions should follow the same format as the other functions. I believe it's clarified in CONTRIBUTING

mattmc3 replied:
Sorry, can you say more? I don't understand what change you're asking for.

belak replied:
```suggestion
function compstyle_zsh-utils_help {
```

It's a small change, but the function style should match between this and the setup function. I think it's also the same with the other _help functions as well.

mattmc3 replied:
Resolved in rename of zsh-utils to zshzoo.

belak commented on completion/functions/compstyle_ohmy_setup:
```diff
@@ -0,0 +1,100 @@
+## Home:     https://ohmyz.sh
+## Authors:  Oh-My-Zsh contributers (https://github.com/ohmyzsh/ohmyzsh/graphs/contributors)
+## License:  MIT
+## Source:   https://github.com/ohmyzsh/ohmyzsh/blob/master/lib/completion.zsh
```
I *think* this should probably refer to the current commit rather than `master` as it can change and it would be nice to know what version this was copied from.

EDIT: same with the prezto_setup function

mattmc3 replied:
Changed to https://github.com/ohmyzsh/ohmyzsh/blob/97134d1eed19b72a9c4bf90a2effa847221608bf/lib/completion.zsh and pushed a new commit.

belak commented:
Thanks for doing this! Overall it looks pretty solid!

> Some starting points:
> 
> * Do we want to include the shameless ripoffs from other projects? Are we okay with goofy 'catsup' names like `ohmy`, `prez`, and `grmlin`?

I'm OK with OMZ and prezto, though my personal preference would be to keep their original names (oh-my-zsh and prezto). That being said, I understand this may introduce some confusion... I also tend to be a function over form kind of person, so maybe there's a better option. I feel like I'm rambling here - essentially, I'm not decided for sure, but it's low hanging fruit at this point of the PR.

> * Grml is GPL 2. Everything else is MIT. Is including code from it okay with you?

I don't think we can include grml because of licensing, as mentioned in another comment.
 
> * I like using `##?` comments for usage documentation, but some find it too clever. If that's not okay, I can switch back to usage strings or HEREDOCs.

For now, I'd prefer to stick to something like usage strings or HEREDOCs.

> * I added a 'cleanup' system that nukes all ':completion:*' zstyles. I know we talked about not doing cleanup. Thoughts on this implementation?

It *should* work and seems like a good first step, though it may clobber any completion styles set by the OS/Distro. I'm not sure what best practices would be here. This also has the unfortunate side-effect of clobbering the caching settings from the module itself which seems less than ideal.

> * Others?

I haven't taken a specific look at the completion styles themselves - I'll try to do that later. I'm not sure if the provided `zsh-utils` completion style would make the most sense as zsh-utils or as mattmc3 - as an example, I tend to name my prompts after the person maintaining them to avoid conflicts in vision. It's why there's a `belak` prompt but not a `zsh-utils` prompt.

mattmc3 committed:
Remove bad comment section per peer review

mattmc3 committed:
Remove GPL2 grml compstyle per peer review

mattmc3 committed:
Reference specific omz commit per peer review

belak commented:
Just ran a quick test on Debian with an empty config:
```
$ zstyle -L ':completion:*'
zstyle ':completion:*:sudo:*' command-path /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /usr/X11R6/bin
```

So, I guess there are OS specific completions which are sometimes set up.

mattmc3 committed:
Remove cleanup functions per peer review

mattmc3 committed:
Use less clever usage string per PR

mattmc3 committed:
Leave cache zstyles in core completions per PR

mattmc3 committed:
Rename compstyle setups to use orig names per PR

mattmc3 commented:
This is all great feedback. I pushed a bunch of commits to address everything you've said so far I think:
- Renamed the setup files to prezto, ohmyzsh, and zshzoo (I prefer naming to my Zsh github org rather than mattmc3)
- Referenced specific completions commits, not just HEAD
- Favored less clever usage method (strings, not doc comments)
- Removed poorly implemented cleanup functionality
- Removed dupe cache zstyles
- Removed grml GPL 2 code
- Adhered better to contributing guidelines

Let me know what other feedback you have. This is really great. Thanks again!

mattmc3 committed:
Fix help syntax

belak commented:
Sorry for the delay - I've been on a much needed vacation for the last few days. I haven't forgotten about this and hope to get to it soon. I don't anticipate any major changes being needed to get this in when I have the time.

mattmc3 commented:
No worries Kaleb. I have the opposite problem in that I had a ton of free time last week (obviously from all my chatter), and this week it's back to the grind. It's not like we're flying planes here - nothing life or death that can't wait a few days (or weeks or months). Whenever you get the chance is fine with me. I'll be around whenever - no rush. In the mean time, I'm using this feature in my daily driver to make sure there's at least some in-the-wild testing.

belak commented:
Thanks for taking the time on this! It looks good to me! We can continue to make incremental improvements on this over time, but this definitely works well as a first version.
